synapse/storage/user_directory.py

```diff
@@ -185,20 +185,38 @@ def _update_profile_in_user_dir_txn(txn):
                         )
                     )
                 else:
-                    sql = """
-                        UPDATE user_directory_search
-                        SET vector = setweight(to_tsvector('english', ?), 'A')
-                            || setweight(to_tsvector('english', ?), 'D')
-                            || setweight(to_tsvector('english', COALESCE(?, '')), 'B')
-                        WHERE user_id = ?
-                    """
-                    txn.execute(
-                        sql,
-                        (
-                            get_localpart_from_id(user_id), get_domain_from_id(user_id),
-                            display_name, user_id,
+                    if new_entry:
+                        sql = """
+                            INSERT INTO user_directory_search(user_id, vector)
+                            VALUES (?,
+                                setweight(to_tsvector('english', ?), 'A')
+                                || setweight(to_tsvector('english', ?), 'D')
+                                || setweight(to_tsvector('english', COALESCE(?, '')), 'B')
+                            )
+                        """
+                        txn.execute(
+                            sql,
+                            (
+                                user_id, get_localpart_from_id(user_id),
+                                get_domain_from_id(user_id), display_name,
+                            )
+                        )
+                    else:
```

Comment:
Can instead we do:

```python
# TODO: Remove this code after we've bumped the minimum version of postgres to
# always support upserts, so we can get rid of `new_entry` usage
if new_entry is True:
    ....
elif new_entry is False:
    ....
else:
    raise RuntimeError("upsert returned None when 'can_native_upsert' is False")
```

Just as a paranoia check :)

Reply:
done